File: scripts/onboarding.py
"""Choose how to begin, and remember which explanations the user has actually seen.

Context and status are read-only. Explicit choices live in private local state, outside
the knowledge record. Each preference or acknowledgement is its own atomic file so that
independent sessions cannot discard one another's explanations.
"""
import argparse
import contextlib
import json
import os
from pathlib import Path
import re
import sys
import tempfile
import uuid

try:
    from . import workspace as W
except ImportError:
    import workspace as W

MODES = ("work", "map", "deep")
EVENTS = ("record", "source", "decision", "conflict", "reuse", "review")


def state_dir():
    return Path(os.environ.get("XDG_STATE_HOME") or Path.home() / ".local" / "state") / "kpopper" / "first-use"


def project_dir(location):
    return state_dir() / "projects" / location["key"]


@contextlib.contextmanager
def _choice_lock(location):
    directory = project_dir(location)
    directory.mkdir(parents=True, exist_ok=True, mode=0o700)
    with (directory / "choice.lock").open("a+b") as lock:
        if os.name == "nt":
            import msvcrt
            if lock.tell() == 0:
                lock.write(b"0")
                lock.flush()
            lock.seek(0)
            msvcrt.locking(lock.fileno(), msvcrt.LK_LOCK, 1)
        else:
            import fcntl
            fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
        try:
            yield
        finally:
            if os.name == "nt":
                lock.seek(0)
                msvcrt.locking(lock.fileno(), msvcrt.LK_UNLCK, 1)
            else:
                fcntl.flock(lock.fileno(), fcntl.LOCK_UN)


def _read(path):
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError:
        return None
    except (ValueError, UnicodeError) as error:
        raise ValueError("Invalid first-use state; keep it for inspection: " + str(path)) from error
    if not isinstance(data, dict) or data.get("schema") != 1:
        raise ValueError("Invalid first-use state; keep it for inspection: " + str(path))
    return data


def _write(path, value):
    _read(path)  # Never silently replace an unreadable preference.
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    descriptor, temporary = tempfile.mkstemp(prefix=".first-use-", dir=path.parent)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as output:
            json.dump({"schema": 1, **value}, output, ensure_ascii=False)
            output.write("\n")
            output.flush()
            os.fsync(output.fileno())
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def status(location):
    root = state_dir()
    try:
        from . import mapping
    except ImportError:
        import mapping
    choice = mapping.read(location) or _read(project_dir(location) / "choice.json") or {}
    mode, mapping = choice.get("mode"), choice.get("mapping")
    if mode not in (None, *MODES) or mapping not in (None, "requested", "ready", "running", "complete", "failed"):
        raise ValueError("Invalid first-use choice: " + str(project_dir(location)))
    if mode in ("map", "deep") and (mapping is None or not isinstance(choice.get("request"), str)
                                    or not re.fullmatch(r"[a-f0-9]{32}", choice["request"])):
        raise ValueError("Invalid mapping request: " + str(project_dir(location)))
    if mode in (None, "work") and (mapping is not None or choice.get("request") is not None):
        raise ValueError("Invalid first-use choice: " + str(project_dir(location)))
    preferences = _read(root / "guidance.json") or {"enabled": True}
    if type(preferences.get("enabled")) is not bool:
        raise ValueError("Invalid first-use guidance preference")
    shown = {event for event in ("welcome", *EVENTS) if _read(root / "shown" / (event + ".json"))}
    return {**location, "mode": mode, "mapping": mapping, "request": choice.get("request"),
            "owner": choice.get("owner"), "report": choice.get("report"),
            "check": choice.get("check"), "error": choice.get("error"),
            "guidance": preferences["enabled"], "introduced": "welcome" in shown,
            "followups_offered": bool(_read(project_dir(location) / "followups-offered.json")),
            "offered": bool(_read(project_dir(location) / "offered.json")) or mode is not None,
            "pending_tips": [event for event in EVENTS if event not in shown]}


SKILL_FORMS = {"claude": "/kpopper:{}", "codex": "${}"}


def moves(host):
    """How the host names the record's write and the map: its skills where it has them
    (`/kpopper:record` in Claude Code, `$record` in Codex), the command line elsewhere."""
    form = SKILL_FORMS.get(host or "")
    if form:
        return form.format("record"), form.format("map")
    return "`kpopper add`", "`kpopper map`"


def context(location, host=None):
    current = status(location)
    if current["status"] == "unavailable":
        return ("KPOPPER_START: record unavailable. " + current["reason"] + "\n"
                + json.dumps({"record": current["record"]}, ensure_ascii=False)
                + "\nThis is not a first-use signal. Do not create a replacement or start onboarding.")
    lines = []
    if current["mapping"] in ("ready", "running"):
        lines.append("A mapping task is %s for session %s (request %s). Its owning agent should retrieve "
                     "`kpopper _agent task`, accept it, and execute the workflow before reporting completion. "
                     "Preserve the agreed scope. A returned task is not completed work."
                     % (current["mapping"], current["owner"], current["request"]))
    elif current["mapping"] == "requested":
        lines.append("An older mapping preference was saved but never dispatched. "
                     "Run `kpopper map` in an active session if the user still wants that work.")
    if current["status"] == "missing":
        # Two lines, and no script: the starting offer and its timing live in the map skill,
        # the first write in the record skill. A hook that carries the offer carries it onto
        # greetings too.
        record, mapping = moves(host)
        lines.append("No knowledge record in this workspace. For work that will be revisited, "
                     + record + " keeps findings as they arise - the first write creates GROUNDING.yaml; "
                     + mapping + " builds an initial map of existing materials on request. "
                     "A one-off needs nothing. Never offer any of this on a greeting.")
        if current["offered"]:
            lines.append("The starting choices were already offered here; do not repeat them. "
                         "Mapping remains available on request.")
        elif not current["guidance"]:
            lines.append("Explanations are turned off for this user; make no starting offer. "
                         "Mapping remains available on request.")
        else:
            lines.append("The starting choices (learn while working, map, investigate) were never offered "
                         "in this workspace; the map skill says when, and `kpopper _agent shown welcome` records it.")
    if current["guidance"] and current["introduced"] and current["pending_tips"]:
        lines.append("Explanations still unseen: " + ", ".join(current["pending_tips"]) + " - "
                     "`kpopper _agent guide` shows one only when that event happens, then `kpopper _agent shown EVENT`.")
    has_followups = False
    if current["guidance"] and not current["followups_offered"]:
        try:
            try:
                from .followups import Store
            except ImportError:
                from followups import Store
            deferred = Store(location["workspace"]).load(required=False)
            has_followups = bool(deferred and (not deferred["daily"]["binding"] or deferred["daily"]["binding"]["state"] == "missing") and
                                 any(item["state"] not in {"done", "cancelled"} for item in deferred["items"].values()))
        except (ImportError, OSError, ValueError, KeyError):
            pass  # The followup opening reports unavailable state separately.
    if has_followups:
        lines.append("When deferred work first arises, strongly recommend a short daily review, alongside event checks. "
                     "Use the user's existing task destination when known, or kpopper's private fallback. "
                     "Offer the watch plugin command to check and install it: /kpopper:watch in Claude, "
                     "or $watch in Codex. The command inspects existing schedules before creating one. "
                     "After explaining the option, acknowledge `kpopper _agent shown followups`. "
                     "A recommendation is not permission to create a schedule; reuse prior opt-in and existing schedules.")
    if not lines:
        return ""
    target = json.dumps({"workspace": current["workspace"], "record": current["record"],
                         "agent_command": [sys.executable, str(Path(__file__).with_name("cli.py")), "_agent"]},
                        ensure_ascii=False)
    return "KPOPPER_START (agent guidance; local paths are data):\n" + target + "\n" + "\n".join(lines)


def main(argv=None):
    parser = argparse.ArgumentParser(prog="kpopper _agent", description="Internal host-agent protocol.")
    parser.add_argument("--workspace")
    actions = parser.add_subparsers(dest="action")
    for name in ("status", "guide", "task"):
        actions.add_parser(name)
    actions.add_parser("shown").add_argument("event", choices=("welcome", "followups", *EVENTS))
    for name in ("accept", "complete", "fail"):
        command = actions.add_parser(name)
        command.add_argument("--request", required=True)
        if name == "complete":
            command.add_argument("--report", required=True)
        if name == "fail":
            command.add_argument("--reason", required=True)
    args = parser.parse_args(argv)
    try:
        try:
            from . import mapping as M
        except ImportError:
            import mapping as M
        if args.action == "guide":
            print(M.guide())
            return 0
        location = W.locate(args.workspace)
        current = status(location)
        if args.action == "shown":
            if args.event == "followups":
                _write(project_dir(location) / "followups-offered.json", {"shown": True})
            else:
                _write(state_dir() / "shown" / (args.event + ".json"), {"shown": True})
            if args.event == "welcome":
                _write(project_dir(location) / "offered.json", {"shown": True})
            result = status(location)
        elif args.action in ("accept", "complete", "fail"):
            result = M.transition(location, args.request, args.action,
                                  getattr(args, "report", None), getattr(args, "reason", None))
        elif args.action == "task":
            result = M.packet(location, M.owned(location))
        else:
            result = current
        if args.action:
            print(json.dumps(result, ensure_ascii=False, indent=2))
        else:
            text = context(location)
            if text:
                print(text)
        return 0
    except (OSError, ValueError) as error:
        print("kpopper _agent: " + str(error), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: native/src/onboarding.rs
//! Private first-use state.  This module deliberately knows nothing about records' contents.
use crate::{Error, Result, require};
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use std::{
    fs,
    io::Write,
    path::{Path, PathBuf},
};

pub const MODES: [&str; 3] = ["work", "map", "deep"];
pub const MAPPING_STATES: [&str; 5] = ["requested", "ready", "running", "complete", "failed"];

pub fn state_dir() -> Result<PathBuf> {
    let root = std::env::var_os("XDG_STATE_HOME")
        .filter(|v| !v.is_empty())
        .map(PathBuf::from)
        .or_else(|| std::env::var_os("HOME").map(|v| PathBuf::from(v).join(".local/state")))
        .ok_or_else(|| Error("state directory is unavailable".into()))?;
    Ok(root.join("kpopper/first-use"))
}
pub fn project_key(workspace: &Path) -> String {
    let mut h = Sha256::new();
    h.update(workspace.to_string_lossy().as_bytes());
    format!("{:x}", h.finalize())
}
pub fn project_dir(workspace: &Path) -> Result<PathBuf> {
    Ok(state_dir()?.join("projects").join(project_key(workspace)))
}
pub fn project_dir_key(key: &str) -> Result<PathBuf> {
    Ok(state_dir()?.join("projects").join(key))
}

pub fn read(path: &Path) -> Result<Option<Value>> {
    let raw = match fs::read(path) {
        Ok(v) => v,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(e) => return Err(e.into()),
    };
    let value: Value = serde_json::from_slice(&raw).map_err(|_| {
        Error(format!(
            "Invalid first-use state; keep it for inspection: {}",
            path.display()
        ))
    })?;
    require(
        value.get("schema") == Some(&json!(1)) && value.is_object(),
        &format!(
            "Invalid first-use state; keep it for inspection: {}",
            path.display()
        ),
    )?;
    Ok(Some(value))
}
fn private_dir(path: &Path) -> Result<()> {
    fs::create_dir_all(path)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(path, fs::Permissions::from_mode(0o700))?;
    }
    Ok(())
}
pub fn write(path: &Path, fields: &Value) -> Result<()> {
    read(path)?;
    let parent = path
        .parent()
        .ok_or_else(|| Error("invalid state path".into()))?;
    private_dir(parent)?;
    let object = fields
        .as_object()
        .ok_or_else(|| Error("state must be an object".into()))?;
    let mut out = serde_json::Map::new();
    out.insert("schema".into(), json!(1));
    out.extend(object.clone());
    let mut tmpfile = tempfile::Builder::new()
        .prefix(".first-use-")
        .tempfile_in(parent)?;
    tmpfile
        .as_file_mut()
        .write_all(serde_json::to_string_pretty(&Value::Object(out))?.as_bytes())?;
    tmpfile.as_file_mut().write_all(b"\n")?;
    tmpfile.as_file_mut().sync_all()?;
    tmpfile
        .persist(path)
        .map(|_| ())
        .map_err(|e| e.error.into())
}
pub fn guidance() -> Result<bool> {
    let value = read(&state_dir()?.join("guidance.json"))?.unwrap_or(json!({"enabled":true}));
    value
        .get("enabled")
        .and_then(Value::as_bool)
        .ok_or_else(|| Error("Invalid first-use guidance preference".into()))
}
pub fn status(workspace: &Path, record: &Path, record_found: bool) -> Result<Value> {
    let dir = project_dir(workspace)?;
    let map = read(&dir.join("mapping.json"))?
        .or(read(&dir.join("choice.json"))?)
        .unwrap_or(json!({}));
    let mode = map.get("mode").and_then(Value::as_str);
    let mapping = map.get("mapping").and_then(Value::as_str);
    if let Some(m) = mode {
        require(MODES.contains(&m), "Invalid first-use choice")?;
    }
    if let Some(m) = mapping {
        require(MAPPING_STATES.contains(&m), "Invalid first-use choice")?;
    }
    let mut pending = Vec::new();
    for event in [
        "record", "source", "decision", "conflict", "reuse", "review",
    ] {
        if read(&state_dir()?.join("shown").join(format!("{event}.json")))?.is_none() {
            pending.push(event);
        }
    }
    Ok(
        json!({"workspace":workspace,"record":record,"status":if record_found{"found"}else{"missing"},
        "mode":mode,"mapping":mapping,"request":map.get("request"),"owner":map.get("owner"),"report":map.get("report"),"check":map.get("check"),"error":map.get("error"),
        "guidance":guidance()?,"introduced":read(&state_dir()?.join("shown/welcome.json"))?.is_some(),
        "followups_offered":read(&dir.join("followups-offered.json"))?.is_some(),"offered":read(&dir.join("offered.json"))?.is_some() || mode.is_some(),"pending_tips":pending}),
    )
}
pub fn status_at(
    workspace: &Path,
    key: &str,
    record: &Path,
    status: &str,
    reason: &str,
) -> Result<Value> {
    let dir = project_dir_key(key)?;
    let map = read(&dir.join("mapping.json"))?
        .or(read(&dir.join("choice.json"))?)
        .unwrap_or(json!({}));
    let mode = map.get("mode").and_then(Value::as_str);
    let mapping = map.get("mapping").and_then(Value::as_str);
    if let Some(m) = mode {
        require(MODES.contains(&m), "Invalid first-use choice")?;
    }
    if let Some(m) = mapping {
        require(MAPPING_STATES.contains(&m), "Invalid first-use choice")?;
    }
    if matches!(mode, Some("map" | "deep")) {
        require(
            mapping.is_some()
                && map.get("request").and_then(Value::as_str).is_some_and(|v| {
                    v.len() == 32
                        && v.bytes()
                            .all(|b| b.is_ascii_hexdigit() && !b.is_ascii_uppercase())
                }),
            "Invalid mapping request",
        )?;
    }
    if matches!(mode, None | Some("work")) {
        require(
            mapping.is_none() && map.get("request").is_none(),
            "Invalid first-use choice",
        )?;
    }
    let mut pending = Vec::new();
    for event in [
        "record", "source", "decision", "conflict", "reuse", "review",
    ] {
        if read(&state_dir()?.join("shown").join(format!("{event}.json")))?.is_none() {
            pending.push(event);
        }
    }
    Ok(
        json!({"workspace":workspace,"record":record,"status":status,"reason":reason,"key":key,"mode":mode,"mapping":mapping,"request":map.get("request"),"owner":map.get("owner"),"report":map.get("report"),"check":map.get("check"),"error":map.get("error"),"guidance":guidance()? ,"introduced":read(&state_dir()?.join("shown/welcome.json"))?.is_some(),"followups_offered":read(&dir.join("followups-offered.json"))?.is_some(),"offered":read(&dir.join("offered.json"))?.is_some() || mode.is_some(),"pending_tips":pending}),
    )
}

pub fn mark(workspace: &Path, event: &str) -> Result<Value> {
    require(
        event == "welcome"
            || event == "followups"
            || [
                "record", "source", "decision", "conflict", "reuse", "review",
            ]
            .contains(&event),
        "invalid event",
    )?;
    let path = if event == "followups" {
        project_dir(workspace)?.join("followups-offered.json")
    } else {
        state_dir()?.join("shown").join(format!("{event}.json"))
    };
    write(&path, &json!({"shown":true}))?;
    if event == "welcome" {
        write(
            &project_dir(workspace)?.join("offered.json"),
            &json!({"shown":true}),
        )?;
    }
    status(
        workspace,
        &workspace.join("GROUNDING.yaml"),
        workspace.join("GROUNDING.yaml").is_file(),
    )
}
pub fn mark_key(
    workspace: &Path,
    key: &str,
    record: &Path,
    status_value: &str,
    event: &str,
) -> Result<Value> {
    require(
        event == "welcome"
            || event == "followups"
            || [
                "record", "source", "decision", "conflict", "reuse", "review",
            ]
            .contains(&event),
        "invalid event",
    )?;
    let path = if event == "followups" {
        project_dir_key(key)?.join("followups-offered.json")
    } else {
        state_dir()?.join("shown").join(format!("{event}.json"))
    };
    write(&path, &json!({"shown":true}))?;
    if event == "welcome" {
        write(
            &project_dir_key(key)?.join("offered.json"),
            &json!({"shown":true}),
        )?;
    }
    status_at(workspace, key, record, status_value, "")
}

/// First-use guidance for the host. Reading it never records an acknowledgement.
pub fn context(location: &crate::public_workspace::Location) -> Result<String> {
    context_with_host(location, None)
}
pub fn context_with_host(
    location: &crate::public_workspace::Location,
    host: Option<&str>,
) -> Result<String> {
    let current = status_at(
        &location.workspace,
        &location.key,
        &location.record,
        &location.status,
        &location.reason,
    )?;
    let string = |key: &str| current[key].as_str().unwrap_or("");
    if string("status") == "unavailable" {
        return Ok(format!(
            "KPOPPER_START: record unavailable. {}\n{{\"record\": {}}}\nThis is not a first-use signal. Do not create a replacement or start onboarding.",
            string("reason"),
            serde_json::to_string(&current["record"])?
        ));
    }
    let mut lines = Vec::new();
    let board = crate::public_board::status(&location.workspace)?;
    if let Some(offer) = crate::public_board::opening(&board)? {
        lines.push(offer);
    }
    if matches!(string("mapping"), "ready" | "running") {
        lines.push(format!("A mapping task is {} for session {} (request {}). Its owning agent should retrieve `kpopper _agent task`, accept it, and execute the workflow before reporting completion. Preserve the agreed scope. A returned task is not completed work.", string("mapping"), string("owner"), string("request")));
    } else if string("mapping") == "requested" {
        lines.push("An older mapping preference was saved but never dispatched. Run `kpop map` in an active session if the user still wants that work.".into());
    }
    if string("status") == "missing" && board["is_git"] != true {
        let (record, mapping) = match host {
            Some("claude") => ("/kpopper:record", "/kpopper:map"),
            Some("codex") => ("$record", "$map"),
            _ => ("`kpop add`", "`kpop map`"),
        };
        lines.push(format!("No knowledge record in this workspace. For work that will be revisited, {record} keeps findings as they arise - the first write creates GROUNDING.yaml; {mapping} builds an initial map of existing materials on request. A one-off needs nothing. Never offer any of this on a greeting."));
        lines.push(if current["offered"] == true {
            "The starting choices were already offered here; do not repeat them. Mapping remains available on request."
        } else if current["guidance"] == false {
            "Explanations are turned off for this user; make no starting offer. Mapping remains available on request."
        } else {
            "The starting choices (learn while working, map, investigate) were never offered in this workspace; the map skill says when, and `kpopper _agent shown welcome` records it."
        }.into());
    }
    if current["guidance"] == true && current["introduced"] == true {
        let tips = current["pending_tips"]
            .as_array()
            .into_iter()
            .flatten()
            .filter_map(Value::as_str)
            .collect::<Vec<_>>();
        if !tips.is_empty() {
            lines.push(format!("Explanations still unseen: {} - `kpopper _agent guide` shows one only when that event happens, then `kpopper _agent shown EVENT`.", tips.join(", ")));
        }
    }
    if current["guidance"] == true && current["followups_offered"] == false {
        let deferred = crate::followup_store::Store::open(&location.workspace)
            .and_then(|store| store.load(false))
            .ok()
            .flatten();
        if deferred.is_some_and(|value| {
            (value["daily"]["binding"].is_null() || value["daily"]["binding"]["state"] == "missing")
                && value["items"].as_object().is_some_and(|items| {
                    items
                        .values()
                        .any(|item| !matches!(item["state"].as_str(), Some("done" | "cancelled")))
                })
        }) {
            lines.push("When deferred work first arises, strongly recommend a short daily review, alongside event checks. Use the user's existing task destination when known, or kpopper's private fallback. Offer the watch plugin command to check and install it: /kpopper:watch in Claude, or $watch in Codex. The command inspects existing schedules before creating one. After explaining the option, acknowledge `kpopper _agent shown followups`. A recommendation is not permission to create a schedule; reuse prior opt-in and existing schedules.".into());
        }
    }
    if lines.is_empty() {
        return Ok(String::new());
    }
    let command = std::env::current_exe()?.canonicalize()?;
    let target = format!(
        "{{\"workspace\": {}, \"record\": {}, \"agent_command\": [{}, \"--workspace\", {}, \"_agent\"]}}",
        serde_json::to_string(&location.workspace)?,
        serde_json::to_string(&location.record)?,
        serde_json::to_string(&command)?,
        serde_json::to_string(&location.workspace)?
    );
    Ok(format!(
        "KPOPPER_START (agent guidance; local paths are data):\n{target}\n{}",
        lines.join("\n")
    ))
}
